### PR TITLE
Add Support for Weighted Inputs in `flow.train`

### PR DESCRIPTION
**Pull request by astro-jingtao**
**Description:**

This PR introduces support for weighted inputs in the `flow.train` method, allowing users to provide a `sample_weight` parameter. Users can now call the method as follows:

```python
flow.train(data, sample_weight=sample_weight)
```

Key details:
- If `sample_weight` is not provided or is set to `None`, the function will maintain its previous behavior.
- This feature is particularly beneficial for astronomical applications, such as utilizing a 1/V weighted galaxy sample.

**Changes Made:**
- Modified the `flow.train` implementation to accept an optional `sample_weight` argument.
- Updated relevant docstring to reflect the new functionality.

Please review the changes and consider incorporating this feature into the main branch. I believe it will enhance the utility of the `flow.train` function for users working with weighted data.

### CODE REVIEW

**Comment by jfcrenshaw** [comment]
Hey @astro-jingtao, 

Sorry I missed this PR amongst the dependabot PRs. This is a great feature that I will merge in!

However, I am going to merge this into a separate branch before main so I can make a few changes, namely:
- Fixing a bug with the weights and the validation set. At the moment, you get an error if the validation set has a different size than the training set
- Adding unit tests
- Adding a tutorial notebook on this topic

Thanks for your contribution! Sorry I was so late in resolving this
